File: javascript/plugins/doppler.js
/* Copyright (c) 2025 Voxgig Ltd, MIT License */

// A port of typescript/plugins/doppler.ts, which is canonical.

const { SekretoError, envkey, providerplugin } = require('../src/provider/support')
const { checkaddr } = require('../src/provider/addr')
const { fetchjson } = require('./httpjson')

/** Doppler.
 *
 * The whole config is downloaded once - Doppler's own bulk endpoint -
 * and answered from memory, like a remote .env: `api.token` is the
 * `API_TOKEN` entry. A service token is config-scoped, so project and
 * config are only needed with broader tokens. */
function dopplerprovider(options) {
  const opts = options || {}

  let values

  const load = async () => {
    if (undefined !== values) {
      return values
    }

    const addr = (opts.addr || 'https://api.doppler.com').replace(/\/$/, '')
    checkaddr(addr)

    let url = addr + '/v3/configs/config/secrets/download?format=json'
    if (opts.project) {
      url += '&project=' + encodeURIComponent(opts.project)
    }
    if (opts.config) {
      url += '&config=' + encodeURIComponent(opts.config)
    }

    const res = await fetchjson('GET', url, {
      authorization: 'Bearer ' + (opts.token || ''),
    })

    if (200 !== res.status || !res.body || 'object' !== typeof res.body) {
      throw new SekretoError('sekreto: doppler error: ' + res.status)
    }

    values = {}
    for (const [key, value] of Object.entries(res.body)) {
      if (null !== value && undefined !== value) {
        values[key] = String(value)
      }
    }

    return values
  }

  return {
    lookup: async (name) => (await load())[envkey(name)],
    describe: () =>
      'doppler' + (opts.project ? ':' + opts.project + '/' + (opts.config || '') : ''),
  }
}

/** The plugin. Needs HTTPS. */
const doppler = providerplugin('doppler', (spec) => dopplerprovider(spec))

module.exports = { doppler, dopplerprovider }

File: dart/plugins/doppler.dart
// Doppler, as a voxgig/plugin definition.
//
// A PLUGIN, not a built-in: it opens a socket. A `Sekreto` can build the
// `doppler` kind only if the calling project imported this file and passed
// `doppler` in the `plugins` option (docs/design/plugin-providers.md).
//
// A port of typescript/plugins/doppler.ts, which is canonical.

import 'dart:async';

import '../src/addr.dart';
import '../src/json.dart';
import '../src/provider.dart';
import '../src/sekreto.dart';
import '../src/support.dart';

import 'httpjson.dart';

/// Doppler.
///
/// The whole config is downloaded once - Doppler's own bulk endpoint - and
/// answered from memory, like a remote .env: `api.token` is the `API_TOKEN`
/// entry. A service token is config-scoped, so project and config are only
/// needed with broader tokens.
class Doppler extends Provider {
  final String? token;
  final String? project;
  final String? config;
  final String? addr;

  Map<String, String>? _values;

  Doppler({this.token, this.project, this.config, this.addr});

  @override
  FutureOr<String?> lookup(String name) {
    final loaded = _values;
    if (null != loaded) {
      return loaded[envkey(name)];
    }

    final useaddr = trimslash(first([addr, 'https://api.doppler.com']));
    checkaddr(useaddr);

    return _load(useaddr).then((values) => values[envkey(name)]);
  }

  Future<Map<String, String>> _load(String useaddr) async {
    var url = '$useaddr/v3/configs/config/secrets/download?format=json';
    if (null != project && project!.isNotEmpty) {
      url += '&project=${uriescape(project!)}';
    }
    if (null != config && config!.isNotEmpty) {
      url += '&config=${uriescape(config!)}';
    }

    final res = await fetchjson('GET', url,
        headers: {'authorization': 'Bearer ${token ?? ''}'});

    final body = res.body.asobj;
    if (200 != res.status || null == body) {
      throw SekretoError('sekreto: doppler error: ${res.status}');
    }

    final loaded = <String, String>{};
    body.forEach((key, value) {
      // A null value is not a secret; it is a declared name with nothing
      // behind it.
      final text = value.text;
      if (null != text) {
        loaded[key] = text;
      }
    });

    // Only a successful load is remembered, so a failure retries.
    _values = loaded;
    return loaded;
  }

  @override
  String describe() {
    if (null != project && project!.isNotEmpty) {
      return 'doppler:$project/${config ?? ''}';
    }
    return 'doppler';
  }
}

/// The `doppler` provider kind. Pass it in `plugins` to name it in a chain.
final Definition doppler = providerplugin(
  'doppler',
  (spec) => Doppler(
    token: spec.token,
    project: spec.project,
    config: spec.config,
    addr: spec.addr,
  ),
);
